OlderThan & NewerThan configured for Days, Hours and Minutes

sinhaashish opened the pull request:
Previously `OlderThan` and `NewerThan` was configured for day
which has been changed to Days, Hours and Minutes. The filed is used in `copy`, `remove` and `mirror` .

This resolves the [bug](https://github.com/minio/mc/issues/2539)

Praveenrajmani commented on cmd/utils.go:
```diff
@@ -161,15 +162,31 @@ func lineTrunc(content string, maxLen int) string {
+		olderThan, e = time.ParseDuration(olderRef)
+		fatalIf(probe.NewError(e), "Unable to parse olderThan=`"+olderRef+"`.")
+	}
+	fmt.Printf("\n objectAge %#v", objectAge)
```
We can remove the debug log prints

sinhaashish replied:
Done

Praveenrajmani commented on cmd/utils.go:
```diff
@@ -161,15 +162,31 @@ func lineTrunc(content string, maxLen int) string {
+		newerThan, e = time.ParseDuration(newerRef)
+		fatalIf(probe.NewError(e), "Unable to parse olderThan=`"+newerRef+"`.")
+	}
+	fmt.Printf("\n objectAge %#v", objectAge)
```
We can remove the debug log prints

sinhaashish replied:
Done.

Praveenrajmani commented on cmd/cp-main.go:
```diff
@@ -41,13 +41,13 @@ var (
+		cli.StringFlag{
 			Name:  "newer-than",
-			Usage: "copy objects newer than N days",
+			Usage: "copy objects newer than N hours and M minutes",
```
Yes as @nitisht suggested, i feel it would be a bit complex for filtering days. It can be `ddhhmm` in my opinion.

sinhaashish replied:
`--expire, -E "168h"          Set expiry in NN[h|m|s]. ` mc expiry uses hours even though 7 days is the default, so i came up with hours and minutes. Will incorporate the days as well.

Praveenrajmani commented on cmd/rm-main.go:
```diff
@@ -321,6 +321,7 @@ func removeRecursive(url string, isIncomplete bool, isFake bool, olderThan int,
 
 // main for rm command.
 func mainRm(ctx *cli.Context) error {
+	fmt.Println(" ashish")
```
We can remove the debug log prints

sinhaashish replied:
Sorry ,Its a mistake.  . Removed the debug log

kannappanr commented on cmd/mirror-main.go:
```diff
@@ -117,8 +117,8 @@ EXAMPLES:
 
-   3. Only mirror files that are newer than 7 days to Amazon S3 cloud storage.
-      $ {{.HelpName}} --newer-than 7 backup/ s3/archive
+   3. Only mirror files that are newer than 7 daysand 10 hours 30 minutes to Amazon S3 cloud storage.
```
`daysand` to `days and`

sinhaashish replied:
done.

kannappanr commented on cmd/mirror-main.go:
```diff
@@ -138,11 +138,11 @@ EXAMPLES:
 
-   9. Mirror objects newer than 10 days from bucket test to a local folder.
-      $ {{.HelpName}} --newer-than=10 s3/test ~/localfolder
+   9. Mirror objects newer than 10 days and  from bucket test to a local folder.
```
`and  from` to `and from`

sinhaashish replied:
done.

kannappanr commented on cmd/mirror-main.go:
```diff
@@ -71,13 +71,13 @@ var (
+		cli.StringFlag{
 			Name:  "older-than",
-			Usage: "filter object(s) older than N days",
+			Usage: "filter object(s) older than L days M hours and N minutes",
```
Should `Usage: "filter object(s) older than L days M hours and N minutes",` be `Usage: "filter object(s) older than L days, M hours and N minutes",`

kannappanr commented on cmd/rm-main.go:
```diff
@@ -103,11 +99,11 @@ EXAMPLES:
 
-   3. Remove all objects older than '90' days recursively from bucket 'jazz-songs' that match 'louis' prefix.
-      $ {{.HelpName}} --recursive --older-than=90 s3/jazz-songs/louis/
+   3. Remove all objects older than '90' days  recursively from bucket 'jazz-songs' that match 'louis' prefix.
```
extra space here

sinhaashish commented:
@vadmeste @kannappanr @Praveenrajmani  PTAL

Praveenrajmani commented on cmd/utils.go:
```diff
@@ -289,3 +301,197 @@ func isURLContains(srcURL, tgtURL, sep string) bool {
+// decimal numbers, each with optional fraction and a unit suffix,
+// such as "300ms", "-1.5h" or "2h45m".
+// Valid time units are "ns", "us" (or "µs"), "ms", "s", "m", "h", "d".
+func ParseDurationTime(s string) (time.Duration, error) {
```
Is there any specific reason for this copy ? We can directly use the go's standard package `time` like `time.ParseDuration("2h15m10s")` 

If we need such forks.. it should be in `pkg/ioutils`

Praveenrajmani commented on cmd/mirror-main.go:
```diff
@@ -640,7 +640,7 @@ func newMirrorJob(srcURL, dstURL string, isFake, isRemove, isOverwrite, isWatch
 }
 
-// copyBucketPolicies - copy policies from source to dest
+// copyBucketPolicies - copy policies from source to dest!= ""
```
The comment here is not clear

Praveenrajmani commented on docs/zh_CN/minio-client-complete-guide.md:
```diff
@@ -360,7 +360,7 @@ FLAGS:
   --fake			   模拟一个假的删除操作。
   --stdin			   从STDIN中读对象列表。
-  --older-than value               删除N天前的对象（默认是0天）。
+  --older-than value        删除超过L天的对象M小时和N分钟LNM [d | h | m]。（默认是0天）。
```
I think we don't need to modify the Chinese docs for the PR. There is a community person who does this for us :) . AFAIK

sinhaashish commented:
@Praveenrajmani  PTAL

Praveenrajmani commented on pkg/ioutils/ioutils_test.go:
```diff
@@ -41,3 +42,33 @@ func (s *MySuite) TestIoutils(c *C) {
+		// Test 4: Input string contains minutes and days
+		{"3m7d", 1512180000000000, ""},
+		// Test 5: Input string contains unknown unit
+		{"4a3d", 0, "time: unknown unit a in duration 4a3d"},
```
I am not sure whether we should support fractional values like `1.5d`. I see `time.ParseDuration` already supports it .
 
```complex, _ := time.ParseDuration("1.5h10m10s")```

Praveenrajmani replied:
looks like our ParseDurationTime gives inconsistent results for such fractional values. It gives me `90h0m0s` for `1.5d`

sinhaashish replied:
@Praveenrajmani Modified. PTAL

harshavardhana commented:
you need to fix the build issues

```
/home/travis/gopath/src/github.com/minio/mc/pkg/ioutils/format.go:1:1: package comment should be of the form "Package ioutils ..."
/home/travis/gopath/src/github.com/minio/mc/pkg/ioutils/format.go:1:1: package comment should not have leading space
Found 2 lint suggestions; failing.
make: *** [lint] Error 1
The command "make test GOFLAGS="-race"" exited with 2.
```

Praveenrajmani commented on pkg/ioutils/format.go:
```diff
@@ -0,0 +1,218 @@
+/*
```
Travis - lint tests are failing here

kannappanr commented on cmd/cp-main.go:
```diff
@@ -93,11 +93,11 @@ EXAMPLES:
-   5. Copy files older than 7 days from Minio cloud storage to Amazon S3 cloud storage.
-      $ {{.HelpName}} --older-than 7 play/mybucket/burningman2011/ s3/mybucket/
+   5. Copy files older than 7 days and 10 hours from Minio cloud storage to Amazon S3 cloud storage.
+      $ {{.HelpName}} --older-than "7d10h" play/mybucket/burningman2011/ s3/mybucket/
```
Have tested the commands. Seems like you need to include `--recursive` for all the older than and newer than flags. Can you check all the commands and fix it?

kannappanr commented on cmd/cp-main.go:
```diff
@@ -93,11 +93,11 @@ EXAMPLES:
-   6. Copy files newer than 7 days from Minio cloud storage to a local path.
-      $ {{.HelpName}} --newer-than 7 play/mybucket/burningman2011/ ~/latest/
+   6. Copy files newer than 7 days and 10 hours from Minio cloud storage to a local path.
+      $ {{.HelpName}} --newer-than "7d10h" play/mybucket/burningman2011/ ~/latest/
```
Also, it looks like we do not have to have double quotes around the value - `7d10h` seems to work.

kannappanr commented:
@sinhaashish Can you resolve the conflicts?

jaipradeesh commented:
This PR helps me remove a lot of code and still do the same job.

A patch with resolved conflicts can be found here - https://gist.githubusercontent.com/jaipradeesh/c680127e314f35a902ae87d8537f8831/raw/041bbe6362386e56bae98833577e477cdbb9292b/gistfile1.txt

sinhaashish commented:
@kannappanr @Praveenrajmani PTAL

deekoder commented:
can we go ahead with the merge on this?

kannappanr commented on docs/minio-client-complete-guide.md:
```diff
@@ -530,8 +530,8 @@ FLAGS:
   --stdin                       read object names from STDIN
-  --older-than value            remove objects older than N days (default: 0)
-  --newer-than value            remove objects newer than N days (default: 0)
+  --older-than value            Remove objects older than L days M hours and N minutes LNM[d|h|m]. (default: 0)
```
Should this be `L days, M hours and`  ?

sinhaashish replied:
edited

kannappanr commented on docs/minio-client-complete-guide.md:
```diff
@@ -530,8 +530,8 @@ FLAGS:
-  --older-than value            remove objects older than N days (default: 0)
-  --newer-than value            remove objects newer than N days (default: 0)
+  --older-than value            Remove objects older than L days M hours and N minutes LNM[d|h|m]. (default: 0)
+  --newer-than value            Remove objects newer than L days M hours and N minutes LNM[d|h|m]. (default: 0)
```
same comment as above.

sinhaashish replied:
done

kannappanr commented on docs/minio-client-complete-guide.md:
````diff
@@ -566,10 +566,10 @@ Removing `play/mybucket/otherobject.txt`.
 Removing `play/mybucket/myobject.1gig`.
 ```
-*Example: Remove object and output a message only if the object is created older than one day. Otherwise, the command stays quiet and nothing is printed out.*
+*Example: Remove object and output a message only if the object is created older than 1 day 2hours and 30 minutes. Otherwise, the command stays quiet and nothing is printed out.*
````
should be `1 day, 2 hours`

sinhaashish replied:
done

kannappanr commented:
It is better to add this support for `find` command also

sinhaashish commented:
> It is better to add this support for `find` command also

Added `days, hours and minutes` for find command. Changed `mc find --older` to `mc find --older-than`.
@Praveenrajmani @kannappanr  PTAL

kannappanr commented on docs/minio-client-complete-guide.md:
```diff
@@ -531,8 +531,8 @@ FLAGS:
   --stdin                       read object names from STDIN
-  --older-than value            remove objects older than N days (default: 0)
-  --newer-than value            remove objects newer than N days (default: 0)
+  --older-than value            Remove objects older than L days, M hours and N minutes LNM[d|h|m]. (default: 0)
```
`Remove` should be `remove`. Rest of the text starts with a lower case letters.

sinhaashish replied:
Done.

kannappanr commented:
LGTM
